pkg/storage/internal/actions/list_actions.go

```diff
@@ -0,0 +1,217 @@
+package actions
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/logging"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/entity"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk/primitives"
+	"github.com/bsv-blockchain/go-sdk/transaction"
+	"github.com/go-softwarelab/common/pkg/must"
+	"github.com/go-softwarelab/common/pkg/slices"
+	"github.com/go-softwarelab/common/pkg/to"
+)
+
+type listActions struct {
+	logger           *slog.Logger
+	transactionsRepo TransactionsRepo
+	outputsRepo      OutputRepo
+	provenTxRepo     ProvenTxRepo
+}
+
+func newListActions(logger *slog.Logger, transactions TransactionsRepo, outputs OutputRepo, provenTx ProvenTxRepo) *listActions {
+	return &listActions{
+		logger:           logging.Child(logger, "list_actions"),
+		transactionsRepo: transactions,
+		outputsRepo:      outputs,
+		provenTxRepo:     provenTx,
+	}
+}
+
+func (l *listActions) ListActions(ctx context.Context, auth wdk.AuthID, args *wdk.ListActionsArgs) (*wdk.ListActionsResult, error) {
+	filter := l.toFilterParams(*auth.UserID, args)
+
+	if l.transactionsRepo == nil {
+		return nil, fmt.Errorf("transactions repository is not initialized")
+	}
+
+	txs, total, err := l.transactionsRepo.ListAndCountActions(ctx, *auth.UserID, filter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list transactions: %w", err)
+	}
+
+	actions := make([]wdk.WalletAction, len(txs))
+	for i, tx := range txs {
+		actions[i] = wdk.WalletAction{
+			TxID:        tx.GetTxID(),
+			Satoshis:    must.ConvertToUInt64(to.NoLessThan(tx.Satoshis, 0)),
+			Status:      string(tx.Status),
+			IsOutgoing:  tx.IsOutgoing,
+			Description: tx.Description,
+		}
+		if tx.Version != nil {
+			actions[i].Version = *tx.Version
+		}
+		if tx.LockTime != nil {
+			actions[i].LockTime = *tx.LockTime
+		}
+	}
+
+	basketIDSet := map[int]struct{}{}
+	allOutputs := map[uint][]*wdk.TableOutput{}
+
+	for _, tx := range txs {
+		_, outputs, err := l.outputsRepo.FindInputsAndOutputsWithBuckets(ctx, tx.TransactionID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load outputs for tx %d: %w", tx.TransactionID, err)
+		}
+		allOutputs[tx.TransactionID] = outputs
+
+		for _, out := range outputs {
+			basketIDSet[out.GetBasketID()] = struct{}{}
+		}
+	}
+
+	var basketIDs []int
```

Comment:
The length is known so preallocate the slice:
```suggestion
basketIDs := make([]int, 0, len(basketIDSet))
```